variantlib/models/variant.py

```diff
@@ -67,69 +111,63 @@ class VariantMeta(BaseModel):
         pttn_nmspc = VALIDATION_NAMESPACE_REGEX[1:-1]
-        pttn_key = VALIDATION_KEY_REGEX[1:-1]
+        pttn_feature = VALIDATION_FEATURE_REGEX[1:-1]
         pttn_value = VALIDATION_VALUE_REGEX[1:-1]
 
-        pattern = rf"^(?P<namespace>{pttn_nmspc})\s*::\s*(?P<key>{pttn_key})\s*::\s*(?P<value>{pttn_value})$"  # noqa: E501
+        pattern = rf"^(?P<namespace>{pttn_nmspc})\s*::\s*(?P<feature>{pttn_feature})\s*::\s*(?P<value>{pttn_value})$"  # noqa: E501
 
         # Try matching the input string with the regex pattern
         match = re.match(pattern, input_str.strip())
 
         if match is None:
             raise ValidationError(
                 f"Invalid format: {input_str}. "
-                "Expected format: '<namespace> :: <key> :: <value>'"
+                "Expected format: '<namespace> :: <feature> :: <value>'"
             )
 
-        # Extract the namespace, key, and value from the match groups
+        # Extract the namespace, feature, and value from the match groups
         namespace = match.group("namespace")
-        key = match.group("key")
+        feature = match.group("feature")
         value = match.group("value")
 
-        # Return an instance of VariantMeta using the parsed values
-        return cls(namespace=namespace, key=key, value=value)
-
-    def serialize(self) -> dict[str, str]:
-        return asdict(self)
-
-    @classmethod
-    def deserialize(cls, data: dict[str, str]) -> Self:
-        assert all(key in data for key in ["namespace", "key", "value"])
-        return cls(**data)
+        # Return an instance of VariantProperty using the parsed values
+        return cls(namespace=namespace, feature=feature, value=value)
 
 
 @dataclass(frozen=True)
 class VariantDescription(BaseModel):
     """
-    A `Variant` is being described by a N >= 1 `VariantMeta` metadata.
-    Each informing the packaging toolkit about a unique `namespace-key-value`
+    A `Variant` is being described by a N >= 1 `VariantProperty`.
+    Each informing the packaging toolkit about a unique `namespace-feature-value`
     combination.
 
     All together they identify the package producing a "variant hash", unique
-    to the exact combination of `VariantMeta` provided for a given package.
+    to the exact combination of `VariantProperty` provided for a given package.
     """
 
-    data: list[VariantMeta] = field(
+    properties: list[VariantProperty] = field(
         metadata={
             "validator": lambda val: validate_and(
                 [
                     lambda v: validate_instance_of(v, list),
-                    lambda v: validate_list_of(v, VariantMeta),
+                    lambda v: validate_list_of(v, VariantProperty),
                     lambda v: validate_list_min_len(v, 1),
-                    lambda v: validate_list_all_unique(v, key=hash),
+                    lambda v: validate_list_all_unique(
+                        v, key=attrgetter("feature_hash")
```

Comment:
I think:

```suggestion
                        v, key=lambda x: (x.namespace, x.key)
```

should work just the same, and avoid unnecessary inheritance. Also, it matches 1:1 what's done below.

Reply:
Oh interesting. I thought I was using this more :D

Reply:
Yes in this PR it's easy. Unfortunately I extensively need and use that concept for sorting & filtering. So if you don't mind I'd like to keep it :D

Reply:
Now I'm confused. Sorting by `(x.namespace, x.key)` sorts by strings, whereas sorting by hash makes sort order pretty random.

Reply:
This is not the sorting part - this is the unicity validation.
This part is the sorting: https://github.com/wheelnext/variantlib/pull/24/files#diff-2739a87bc9b043288bb679e59c1a2200798497f3d166f9e8fdea8d0620cce6d7R188

Reply:
> Now I'm confused. Sorting by (x.namespace, x.key) sorts by strings, whereas sorting by hash makes sort order pretty random.

For the record. Yes as long as you would use the `hash()` function inside python yes. If you were to use a real hashing function - that would not be the case ;) It would be random but consistent. And only the consistency matters not the order. The problem with `hash()` is that it's seeded at random and not consistent.

But anyway doesn't matter since we are not doing that ;)